# Análise de uma Matriz de Insumo-Produto Brasileira

---

## Conceitos 

- **Matriz Insumo-Produto:** um modelo que representa os setores da economia de um país ou de uma região através de uma matriz, para prever o impacto de alterações de um setor sobre os outros e de consumidores, do governo e de fornecedores estrangeiros sobre a economia, desenvolvido por [Wassily Leontief](https://pt.wikipedia.org/wiki/Wassily_Leontief). Cada coluna da matriz representa o valor monetário de insumos (*inputs*) de um setor e cada linha representa o valor monetário das saídas (*outputs*) de um setor, mostrando as relações de dependência dos setores da economia. 


- **Matriz $A$:** Matriz que relaciona os valores de cada posição na matriz ($z_{ij}$) com a produção total do setor $i$. Matematicamente temos: 

$$x_i \equiv \sum_{j=1}^n z_{ij} + y_i \forall i,j = 1, 2, ..., n$$

$$a_{ij} = \frac{z_{ij}}{x_i} \forall i,j = 1, 2, ..., n$$

Juntando as duas equações podemos reescrever da seguinte forma:

$$x_i = \sum_{j=1}^n a_{ij}x_j + y_i$$

Em termos matriciais temos:

$$x = Ax + y$$

$$\begin{bmatrix}
x_1\\
x_2\\
...\\
x_i\\
...\\
x_n
\end{bmatrix} =
\begin{bmatrix}
a_{11} & a_{12} & ... & a_{1i} & ... & a_{1n}\\
a_{21} & a_{22} & ... & a_{2i} & ... & a_{2n}\\
...    & ...    & ... & ...    & ... & ...\\
a_{i1} & a_{i2} & ... & a_{ii} & ... & a_{in}\\
...    & ...    & ... & ...    & ... & ...\\
a_{n1} & a_{n2} & ... & a_{ni} & ... & a_{nn}
\end{bmatrix}
\begin{bmatrix}
x_1\\
x_2\\
...\\
x_i\\
...\\
x_n
\end{bmatrix} +
\begin{bmatrix}
y_1\\
y_2\\
...\\
y_i\\
...\\
y_n
\end{bmatrix}$$

Assim podemos fazer manipulações algébricas:

$$x - Ax = y$$
$$(I - A)x = y$$
$$x = (I - A)^{-1}y$$

Sendo $x$ o produto final ($VBP$), e $(I - A)^{-1}$ a matriz B, ou matriz inversa de Leontief, que mostra os requisitos totais, diretos e indiretos, para a produção de cada setor. Podemos pensar nisso como uma análise de variação. 

$$\Delta x = (I - A)^{-1}\Delta y$$

Ou seja, uma variação na demanda final ($\Delta y$) causa um aumento na produção ($\Delta x$) de acordo com a tecnologia $(I - A)^{-1}$. Assim, assume-se que a economia é impulsionada por variações exógenas dado as relações endógenas.

- **Demanda Intermediária ($DI$):** É a soma das demandas internas dos setores da economia, consumida pelos setores. A produção total de setor, isto é, a soma dos valores de uma linha, equivale à Demanda Intermediária daquele setor.


- **Demanda Final ($DF$ ou $y_i$):** Demanda dos consumidores finais (famílias, governo, exportações e investimentos), exlui-se o que é consumido para confecção de novos produtos (a demanda intermediária).


- **Consumo Intermediário: ($CI$)** É o consumo total de insumos de cada setor. Em outras palavras, equivale à soma dos valores de uma determinada coluna.

- **Valor Adicionado da Produção ($VAB$):** É o valor que cada setor da economia acresce ao valor final de tudo que foi produzido em uma região. Uma das formas de calcular o Produto Interno Bruto ($PIB$) de uma economia é pela soma dos VABs setoriais e dos impostos. É a principal medida do tamanho total de uma economia.


- **Valor Bruto da Produção ($VBP$):** O Valor Bruto da Produção equivale à soma da produção total ($DI$) com a demanda final: $VBP = DI + DF$. Também pode ser calculado pela soma do Consumo Intermediário com o Valor Adicionado da Produção de cada setor: $VBP = CI + VAB$. Com isso, temos que a soma das demandas equivale à soma das produções: $DI + DF = CI + VAB$.


- **Encadeamentos para a frente (sensibilidade da dispersão):** relações de um setor B com outros setores como fornecedor.


- **Encadeamentos para trás (poder de dispersão):** relações de um setor C com outros setores como demandante.

## 1. Definição do Problema
---

A Matriz de Insumo-Produto é disponibilizada numa tabela em formato *.xlsx*, porém os dados estão desalinhados, o que impossibilita o uso direto da tabela para análises de ligações. 

## 2. Obtenção dos dados
---

O primeiro passo é baixar os dados do IBGE. Para isso, acessamos [este site](https://www.ibge.gov.br/estatisticas/economicas/contas-nacionais/9085-matriz-de-insumo-produto.html?=&t=resultados), e a partir dele podemos retirar o link para download da Matriz de Insumo-Produto Nível 67. 

**Mas o que significa Nível 67?** Significa que a economia está dividida em 67 setores, desde agricultura até serviços domésticos.

Para baixarmos os dados diretos do site, utilizamos a biblioteca ``urllib``:

In [1]:
from urllib.request import urlretrieve

url = 'ftp://ftp.ibge.gov.br/Contas_Nacionais/Matriz_de_Insumo_Produto/2015/Matriz_de_Insumo_Produto_2015_Nivel_67.xls'

urlretrieve(url,
            filename='Matrizes/matriz_insumo_produto_67.xls')

('Matrizes/matriz_insumo_produto_67.xls',
 <email.message.Message at 0x1cac3fefec8>)

## 3. Exploração dos dados
---
Agora temos os dados "crus". Estes dados mostram o valor monetário da produção de cada um dos setores, e em que parte da cadeia produtiva ele se encontra. Usaremos ``pandas`` e ``numpy`` para manipulação, exploração e análise dos dados.

In [2]:
import pandas as pd
import numpy as np

In [4]:
# Este arquivo possui múltiplas planilhas, então precisamos especificar qual nos interessa
# Neste caso, queremos a tabela de oferta e demanda da produção nacional, que é a planilha 03
matriz = pd.read_excel("Matrizes/matriz_insumo_produto_67.xls",
                       sheet_name="03", 
                       skiprows=3)
matriz.head(5)

Unnamed: 0                                         Unnamed: 1  \
0        NaN                                                NaN   
1      01911                      Arroz, trigo e outros cereais   
2      01912                                      Milho em grão   
3      01913  Algodão herbáceo, outras fibras da lavoura tem...   
4      01914                                     Cana-de-açúcar   

   Valor da produção  \
0                NaN   
1            11036.0   
2            29975.0   
3             8943.0   
4            46080.0   

   0191\nAgricultura, inclusive o apoio à agricultura e a pós-colheita  \
0                                                NaN                     
1                                              166.0                     
2                                              566.0                     
3                                              260.0                     
4                                              615.0                     

   0192\nPecuária, inclusive o apoio à pecuária  \
0                                           NaN   
1                                         164.0   
2                                        2314.0   
3                                          23.0   
4                                         148.0   

   0280\nProdução florestal; pesca e aquicultura  \
0                                            NaN   
1                                            0.0   
2                                            9.0   
3                                            0.0   
4                                            4.0   

   0580\nExtração de carvão mineral e de minerais não metálicos  \
0                                                NaN              
1                                                0.0              
2                                                0.0              
3                                                0.0              
4                                                0.0              

   0680\nExtração de petróleo e gás, inclusive as atividades de apoio  \
0                                                NaN                    
1                                                0.0                    
2                                                0.0                    
3                                                0.0                    
4                                                0.0                    

   0791\nExtração de minério de ferro, inclusive beneficiamentos e a aglomeração  \
0                                                NaN                               
1                                                0.0                               
2                                                0.0                               
3                                                0.0                               
4                                                0.0                               

   0792\nExtração de minerais metálicos não ferrosos, inclusive beneficiamentos  \
0                                                NaN                              
1                                                0.0                              
2                                                0.0                              
3                                                0.0                              
4                                                0.0                              

   ...  9700\nServiços domésticos  Total\ndo produto  \
0  ...                        NaN                NaN   
1  ...                        0.0             8843.0   
2  ...                        0.0            11179.0   
3  ...                        0.0             3722.0   
4  ...                        0.0            44764.0   

   Exportação\nde bens e\nserviços  Consumo\ndo governo  \
0                              NaN                  NaN   
1                           1280.0                  0.0   
2                          14

In [5]:
matriz.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 79 columns):
 #   Column                                                                             Non-Null Count  Dtype  
---  ------                                                                             --------------  -----  
 0   Unnamed: 0                                                                         129 non-null    object 
 1   Unnamed: 1                                                                         127 non-null    object 
 2   Valor da produção                                                                  128 non-null    float64
 3   0191
Agricultura, inclusive o apoio à agricultura e a pós-colheita                 128 non-null    float64
 4   0192
Pecuária, inclusive o apoio à pecuária                                        128 non-null    float64
 5   0280
Produção florestal; pesca e aquicultura                                       128 non-null    float64

In [6]:
print(f'Tamanho da matriz: {matriz.shape[0]} linhas x {matriz.shape[1]} colunas')
if matriz.shape[0] != matriz.shape[1]:
    print("Deveria ser quadrada!")

Tamanho da matriz: 132 linhas x 79 colunas
Deveria ser quadrada!


In [7]:
for coluna in matriz.columns:
    print(f'NaNs na coluna {coluna}: {matriz.loc[:, coluna].isnull().sum()} valores')

NaNs na coluna Unnamed: 0: 3 valores
NaNs na coluna Unnamed: 1: 5 valores
NaNs na coluna Valor da produção: 4 valores
NaNs na coluna 0191
Agricultura, inclusive o apoio à agricultura e a pós-colheita: 4 valores
NaNs na coluna 0192
Pecuária, inclusive o apoio à pecuária: 4 valores
NaNs na coluna 0280
Produção florestal; pesca e aquicultura: 4 valores
NaNs na coluna 0580
Extração de carvão mineral e de minerais não metálicos: 4 valores
NaNs na coluna 0680
Extração de petróleo e gás, inclusive as atividades de apoio: 4 valores
NaNs na coluna 0791
Extração de minério de ferro, inclusive beneficiamentos e a aglomeração: 4 valores
NaNs na coluna 0792
Extração de minerais metálicos não ferrosos, inclusive beneficiamentos: 4 valores
NaNs na coluna 1091
Abate e produtos de carne, inclusive os produtos do laticínio e da pesca: 4 valores
NaNs na coluna 1092
Fabricação e refino de açúcar: 4 valores
NaNs na coluna 1093
Outros produtos alimentares: 4 valores
NaNs na coluna 1100
Fabricação de bebidas

## 4. Tratamento dos dados

### Problemas a serem resolvidos: 

- Células com texto e valores ``NaN``;

- A maior parte dos nomes das colunas têm números e `\n`;

- Os setores estão dividos nas linhas mas nas colunas não (ou seja, não é uma matriz quadrada, o que impede a sua inversão)

Primeiro, apagamos as linhas desnecessárias

In [8]:
matriz.drop(index=[0, 128, 130, 131], inplace=True)
matriz

Unnamed: 0                                         Unnamed: 1  \
1        01911                      Arroz, trigo e outros cereais   
2        01912                                      Milho em grão   
3        01913  Algodão herbáceo, outras fibras da lavoura tem...   
4        01914                                     Cana-de-açúcar   
5        01915                                      Soja  em grão   
..         ...                                                ...   
124      94801  Organizações patronais, sindicais e outros ser...   
125      94802  Manutenção de computadores, telefones e objeto...   
126      94803                                  Serviços pessoais   
127      97001                                Serviços domésticos   
129      Total                                                NaN   

     Valor da produção  \
1              11036.0   
2              29975.0   
3               8943.0   
4              46080.0   
5             109170.0   
..                 ...   
124            83334.0   
125            26429.0   
126            42366.0   
127            61996.0   
129         10226869.0   

     0191\nAgricultura, inclusive o apoio à agricultura e a pós-colheita  \
1                                                166.0                     
2                                                566.0                     
3                                                260.0                     
4                                                615.0                     
5                                               1864.0                     
..                                                 ...                     
124                                              104.0                     
125                                                0.0                     
126                                                0.0                     
127                                                0.0                     
129                                           115679.0                     

     0192\nPecuária, inclusive o apoio à pecuária  \
1                                           164.0   
2                                          2314.0   
3                                            23.0   
4                                           148.0   
5                                           155.0   
..                                            ...   
124                                          29.0   
125                                           0.0   
126                                           0.0   
127                                           0.0   
129                                       55512.0   

     0280\nProdução florestal; pesca e aquicultura  \
1                                              0.0   
2                                              9.0   
3                                              0.0   
4                                              4.0   
5                                              0.0   
..                                             ...   
124                                           14.0   
125                                            5.0   
126                                            0.0   
127                                            0.0   
129                                         6546.0   

     0580\nExtração de carvão mineral e de minerais não metálicos  \
1                                                  0.0              
2                                                  0.0              
3                                                  0.0              
4                                                  0.0              
5                                                  0.0              
..                                                 ...              
124                                               35.0              
125                                                0.0              
126                                       

O setor "*0191 Agricultura, inclusive o apoio à agricultura e o pós-colheita*" está dividido em várias partes, mas todas possuem um código começando com 0191!

Isso permite a automatização da soma das divisões setoriais.

In [9]:
colunas = list(matriz.columns)

matriz_insumo_produto = pd.DataFrame(columns=matriz.columns)

for i in range(len(colunas)):
    indices = []
    
    for j in range(1, len(matriz.index)):
        
        subsetor = matriz.loc[j, 'Unnamed: 0']
        
        if subsetor[:4] in list(matriz.columns)[i]:
            # se o começo do código bate com o código desta coluna, marque-a para ser adicionada
            indices.append(matriz.index[j-1])
            
    if len(indices) != 0:
        # se há colunas para serem somadas, as some e as adicione na matriz_insumo_produto
        matriz_insumo_produto = matriz_insumo_produto.append(
            matriz.loc[indices[0]:indices[-1]].sum(),
            ignore_index=True
        )
        
matriz_insumo_produto

Unnamed: 0  \
0   019110191201913019140191501916019170191801919   
1                            01921019220192301924   
2                                      0280102802   
3                                      0580105802   
4                                           06801   
..                                            ...   
62                                          86911   
63                                          86921   
64                                          90801   
65                                948019480294803   
66                                          97001   

                                           Unnamed: 1  Valor da produção  \
0   Arroz, trigo e outros cereaisMilho em grãoAlgo...           302321.0   
1   Bovinos e outros animais vivos, produtos anima...           128920.0   
2   Produtos da exploração florestal e da silvicul...            34101.0   
3                Carvão mineralMinerais não metálicos            20365.0   
4           Petróleo, gás natural e serviços de apoio           162130.0   
..                                                ...                ...   
62                                      Saúde pública           175278.0   
63                                      Saúde privada           224955.0   
64    Serviços de artes, cultura, esporte e recreação            33796.0   
65  Organizações patronais, sindicais e outros ser...           152129.0   
66                                Serviços domésticos            61996.0   

    0191\nAgricultura, inclusive o apoio à agricultura e a pós-colheita  \
0                                              6657.0                     
1                                               521.0                     
2                                               915.0                     
3                                                46.0                     
4                                                 0.0                     
..                                                ...                     
62                                                0.0                     
63                                                0.0                     
64                                                0.0                     
65                                              104.0                     
66                                                0.0                     

    0192\nPecuária, inclusive o apoio à pecuária  \
0                                         3635.0   
1                                         4475.0   
2                                         1095.0   
3                                          356.0   
4                                            0.0   
..                                           ...   
62                                           0.0   
63                                           0.0   
64                                           0.0   
65                                          29.0   
66                                           0.0   

    0280\nProdução florestal; pesca e aquicultura  \
0                                            78.0   
1                                            45.0   
2                                          1728.0   
3                                            11.0   
4                                             0.0   
..                                            ...   
62                                            0.0   
63                                            0.0   
64                                            0.0   
65                                           19.0   
66                                            0.0   

    0580\nExtração de carvão mineral e de minerais não metálicos  \
0                                                 0.0              
1                                                 0.0              
2                                                 2.0              
3                                               324.0 

Agora corrigimos os nomes das colunas. Uma forma interessante de fazer isso é usando ***regex*** (expressões regulares). Para isso, importamos o módulo `re`:

In [10]:
import re

colunas = list(matriz_insumo_produto.columns)
for i in range(len(colunas)):
    colunas[i] = re.sub('\d', "", colunas[i])  # Se há algum número no nome da coluna, o substitua por nada ("")
    colunas[i] = re.sub('\n', " ", colunas[i])  # Se há algum "\n" (pular para nova linha) o substituia por um espaço (" ")
    colunas[i] = colunas[i].strip()  # retira espaços em branco no começo e final da string
        
matriz_insumo_produto.columns = colunas

- VBP: Valor Bruto da Produção

- DF: Demanda Final

As variáveis estão escritas em caixa alta pois se tratam de valores constantes, seguindo a formatação recomendada na PEP 8.

In [11]:
VBP = matriz_insumo_produto["Demanda total"].copy()
DF = matriz_insumo_produto["Demanda final"].copy()

Retiramos agora as três primeiras colunas, pois não precisamos delas, e as duas últimas colunas, pois já guardamos estas informações em outras variáveis.

``inplace=True`` aplica o método `.drop()` na própria matriz

In [11]:
matriz_insumo_produto.drop(columns=(colunas[:3] + colunas[70:]),
                           axis=1,
                           inplace=True)

matriz_insumo_produto

Agricultura, inclusive o apoio à agricultura e a pós-colheita  \
0                                              6657.0               
1                                               521.0               
2                                               915.0               
3                                                46.0               
4                                                 0.0               
..                                                ...               
62                                                0.0               
63                                                0.0               
64                                                0.0               
65                                              104.0               
66                                                0.0               

    Pecuária, inclusive o apoio à pecuária  \
0                                   3635.0   
1                                   4475.0   
2                                   1095.0   
3                                    356.0   
4                                      0.0   
..                                     ...   
62                                     0.0   
63                                     0.0   
64                                     0.0   
65                                    29.0   
66                                     0.0   

    Produção florestal; pesca e aquicultura  \
0                                      78.0   
1                                      45.0   
2                                    1728.0   
3                                      11.0   
4                                       0.0   
..                                      ...   
62                                      0.0   
63                                      0.0   
64                                      0.0   
65                                     19.0   
66                                      0.0   

    Extração de carvão mineral e de minerais não metálicos  \
0                                                 0.0        
1                                                 0.0        
2                                                 2.0        
3                                               324.0        
4                                                 1.0        
..                                                ...        
62                                                0.0        
63                                                0.0        
64                                                0.0        
65                                               35.0        
66                                                0.0        

    Extração de petróleo e gás, inclusive as atividades de apoio  \
0                                                 0.0              
1                                                 0.0              
2                                                 0.0              
3                                               534.0              
4                                             12079.0              
..                                                ...              
62                                                0.0              
63                                                0.0              
64                                                0.0              
65                                              531.0              
66                                                0.0              

    Extração de minério de ferro, inclusive beneficiamentos e a aglomeração  \
0                                                 0.0                         
1                                                 0.0                         
2                                                 0.0                         
3                                                 0.0                         
4                                                46.0                         
..                   

Ao deixarmos todos os índices com o mesmo nome evitamos erros de processamento

In [12]:
matriz_insumo_produto.index = matriz_insumo_produto.columns
VBP.index = matriz_insumo_produto.columns
DF.index = matriz_insumo_produto.columns

Resolvido! Agora temos nossa matriz de insumo-produto pronta para ser manipulada!

In [13]:
import pymrio

Agora, importamos um pacote específico para análises de matriz de Insumo-Produto: o *pymrio*.

*Pymrio: Multi-Regional Input-Output Analysis in Python - A python module for automating io calculations and generating reports*

In [14]:
matriz_A = pymrio.calc_A(matriz_insumo_produto, VBP)
matriz_A

Agricultura, inclusive o apoio à agricultura e a pós-colheita  \
Agricultura, inclusive o apoio à agricultura e ...                                           0.022020               
Pecuária, inclusive o apoio à pecuária                                                       0.001723               
Produção florestal; pesca e aquicultura                                                      0.003027               
Extração de carvão mineral e de minerais não me...                                           0.000152               
Extração de petróleo e gás, inclusive as ativid...                                           0.000000               
...                                                                                               ...               
Saúde pública                                                                                0.000000               
Saúde privada                                                                                0.000000               
Atividades artísticas, criativas e de espetáculos                                            0.000000               
Organizações associativas e outros serviços pes...                                           0.000344               
Serviços domésticos                                                                          0.000000               

                                                    Pecuária, inclusive o apoio à pecuária  \
Agricultura, inclusive o apoio à agricultura e ...                                0.028196   
Pecuária, inclusive o apoio à pecuária                                            0.034711   
Produção florestal; pesca e aquicultura                                           0.008494   
Extração de carvão mineral e de minerais não me...                                0.002761   
Extração de petróleo e gás, inclusive as ativid...                                0.000000   
...                                                                                    ...   
Saúde pública                                                                     0.000000   
Saúde privada                                                                     0.000000   
Atividades artísticas, criativas e de espetáculos                                 0.000000   
Organizações associativas e outros serviços pes...                                0.000225   
Serviços domésticos                                                               0.000000   

                                                    Produção florestal; pesca e aquicultura  \
Agricultura, inclusive o apoio à agricultura e ...                                 0.002287   
Pecuária, inclusive o apoio à pecuária                                             0.001320   
Produção florestal; pesca e aquicultura                                            0.050673   
Extração de carvão mineral e de minerais não me...                                 0.000323   
Extração de petróleo e gás, inclusive as ativid...                                 0.000000   
...                                                                                     ...   
Saúde pública                                                                      0.000000   
Saúde privada                                                                      0.000000   
Atividades artísticas, criativas e de espetáculos                                  0.000000   
Organizações associativas e outros serviços pes...                                 0.000557   
Serviços domésticos                                                                0.000000   

                                                    Extração de carvão mineral e de minerais não metálicos  \
Agricultura, inclusive o apoio à agricultura e ...                                           0.000000        
Pecuária, inclusive o apoio à pecuária                                                       0.000000        
Produção florestal; pesca e aquicultura        

In [15]:
matriz_L = pymrio.calc_L(matriz_A)
matriz_L

Agricultura, inclusive o apoio à agricultura e a pós-colheita  \
Agricultura, inclusive o apoio à agricultura e ...                                           1.027523               
Pecuária, inclusive o apoio à pecuária                                                       0.002005               
Produção florestal; pesca e aquicultura                                                      0.004008               
Extração de carvão mineral e de minerais não me...                                           0.002909               
Extração de petróleo e gás, inclusive as ativid...                                           0.023635               
...                                                                                               ...               
Saúde pública                                                                                0.000000               
Saúde privada                                                                                0.000000               
Atividades artísticas, criativas e de espetáculos                                            0.000145               
Organizações associativas e outros serviços pes...                                           0.001689               
Serviços domésticos                                                                          0.000000               

                                                    Pecuária, inclusive o apoio à pecuária  \
Agricultura, inclusive o apoio à agricultura e ...                                0.056085   
Pecuária, inclusive o apoio à pecuária                                            1.041045   
Produção florestal; pesca e aquicultura                                           0.010536   
Extração de carvão mineral e de minerais não me...                                0.004904   
Extração de petróleo e gás, inclusive as ativid...                                0.020266   
...                                                                                    ...   
Saúde pública                                                                     0.000000   
Saúde privada                                                                     0.000000   
Atividades artísticas, criativas e de espetáculos                                 0.000270   
Organizações associativas e outros serviços pes...                                0.001780   
Serviços domésticos                                                               0.000000   

                                                    Produção florestal; pesca e aquicultura  \
Agricultura, inclusive o apoio à agricultura e ...                                 0.010155   
Pecuária, inclusive o apoio à pecuária                                             0.001761   
Produção florestal; pesca e aquicultura                                            1.053728   
Extração de carvão mineral e de minerais não me...                                 0.000822   
Extração de petróleo e gás, inclusive as ativid...                                 0.010180   
...                                                                                     ...   
Saúde pública                                                                      0.000000   
Saúde privada                                                                      0.000000   
Atividades artísticas, criativas e de espetáculos                                  0.000138   
Organizações associativas e outros serviços pes...                                 0.001250   
Serviços domésticos                                                                0.000000   

                                                    Extração de carvão mineral e de minerais não metálicos  \
Agricultura, inclusive o apoio à agricultura e ...                                           0.004839        
Pecuária, inclusive o apoio à pecuária                                                       0.000162        
Produção florestal; pesca e aquicultura        

Chegamos enfim à matriz de coeficientes!

Agora podemos exportar estas matrizes para arquivos csv, para evitar calcularmos novamente estes valores

In [16]:
import os

pasta = "Matrizes"

if not os.path.exists(pasta):
    # garante que a pasta onde quero salvar estes arquivos existe
    os.mkdir(pasta)

matriz_insumo_produto.to_csv("Matrizes\matriz_insumo_produto.csv")
matriz_A.to_csv("Matrizes\matriz_A.csv")
matriz_L.to_excel("Matrizes\matriz_coeficientes.xlsx")
VBP.to_csv("Matrizes\VBP.csv", header=True)
DF.to_csv("Matrizes\demanda_final.csv", header=True)

## 5. Análise dos dados

---

Com a matriz de coeficientes pronta, podemos realizar as análises e descobrir o tipo dos setores

Os tipos dependem do coeficiente das ligações para frente e para trás. Calculamos isso da seguinte forma:

$$L_f = \frac{\frac{1}{n}\sum_{j}{b_{ij}}}{\frac{1}{n^2}\sum_i\sum_j b_{ij}}$$


$$L_t = \frac{\frac{1}{n}\sum_{i}{b_{ij}}}{\frac{1}{n^2}\sum_i\sum_j b_{ij}}$$

Ou seja: o coeficiente de ligação para frente da linha $i$ é igual à média dos coeficientes desta linha dividida pela média de todos os coeficientes da matriz;

E o coeficiente de ligações para trás da coluna $j$ é igual à média dos coeficientes desta coluna dividida pela média total dos coeficientes.

Eu realizo a transformação para uma matriz com `.to_numpy()` antes de tirar a média de todos os valores pois é uma operação muito mais eficiente. O método `.mean()` em pandas apenas realiza a média de todas as linhas, então precisaríamos realizar a média duas vezes.

In [17]:
media = matriz_L.to_numpy(copy=True).mean()

l_f = matriz_L.mean(axis=0) / media # axis=0 para média das colunas
l_t = matriz_L.mean(axis=1) / media # axis=1 para média das colunas

In [18]:
df_coeficientes = pd.DataFrame({
    "Setor": list(matriz_L.columns),
    "Ligações para frente": l_f.to_numpy(),
    "Ligações para trás": l_t.to_numpy()
})
df_coeficientes

Setor  Ligações para frente  \
0   Agricultura, inclusive o apoio à agricultura e...              0.965247   
1              Pecuária, inclusive o apoio à pecuária              1.018823   
2             Produção florestal; pesca e aquicultura              0.742157   
3   Extração de carvão mineral e de minerais não m...              0.995418   
4   Extração de petróleo e gás, inclusive as ativi...              0.963418   
..                                                ...                   ...   
62                                      Saúde pública              0.814404   
63                                      Saúde privada              0.840359   
64  Atividades artísticas, criativas e de espetáculos              0.873071   
65  Organizações associativas e outros serviços pe...              0.902033   
66                                Serviços domésticos              0.553941   

    Ligações para trás  
0             1.862446  
1             0.870680  
2             0.762387  
3             0.688005  
4             1.406660  
..                 ...  
62            0.553941  
63            0.611399  
64            0.640074  
65            0.724536  
66            0.553941  

[67 rows x 3 columns]

Podemos dividir os setores em quatro tipos:

- **Tipo I:** ($L_t < 1; L_f < 1$) Geralmente independente, sem ligações ou ligações fracas;

- **Tipo II:** ($L_t < 1; L_f > 1$) Depende da demanda interisdustrial, mais relações para frente;

- **Tipo III:** ($L_t > 1; L_f > 1$) Geralmente dependente, é um ***setor chave***, contribui acima da média para o crescimento da economia;

- **Tipo IV:** ($L_t > 1; L_f < 1$) Depende da oferta interindustrial, mais relações para trás;

In [22]:
tipos = []
chaves = []

for i in range(df_coeficientes.shape[0]):
    chave = ''
    if df_coeficientes.iloc[i, 1] < 1 and df_coeficientes.iloc[i, 2] < 1:
        tipo = 'I'
    elif df_coeficientes.iloc[i, 1] < 1 and df_coeficientes.iloc[i, 2] > 1:
        tipo = 'II'
    elif df_coeficientes.iloc[i, 1] > 1 and df_coeficientes.iloc[i, 2] > 1:
        tipo = 'III'
        chave = '*'
    else:
        tipo = 'IV'
    tipos.append(tipo)
    chaves.append(chave)

df_coeficientes['Tipo'] = tipos
df_coeficientes['Setor Chave'] = chaves
df_coeficientes[df_coeficientes['Setor Chave'] == '*']

Setor  Ligações para frente  \
16  Fabricação de celulose, papel e produtos de papel              1.132302   
18                     Refino de petróleo e coquerias              1.350624   
20  Fabricação de químicos orgânicos e inorgânicos...              1.115801   
21  Fabricação de defensivos, desinfestantes, tint...              1.121799   
24  Fabricação de produtos de borracha e de materi...              1.143552   
26  Produção de ferro gusa/ferroligas, siderurgia ...              1.222849   
28  Fabricação de produtos de metal, exceto máquin...              1.077724   
37  Energia elétrica, gás natural e outras utilidades              1.162158   
41                               Transporte terrestre              1.121909   

    Ligações para trás Tipo Setor Chave  
16            1.008885  III           *  
18            2.823324  III           *  
20            1.909688  III           *  
21            1.134049  III           *  
24            1.202437  III           *  
26            1.043301  III           *  
28            1.053088  III           *  
37            2.027143  III           *  
41            2.251273  III           *

É notável um padrão entre os possíveis setores chave: *a maioria deles é do setor industrial*. Um ponto muito comentado por economistas desenvolvimentistas é a importância da retomada do crescimento deste setor no Brasil para alavancar a economia.

Uma relação interessante que podemos visualizar é a seguinte:

$$L \cdot DF = VBP$$

Isto é, a multiplicação da matriz de coeficientes pelo vetor da Demanda Final resulta no VBP! Com isso, podemos prever qual seria a variação do VBP ao mudarmos cada valor no DF.

Vejamos, qual seria o impactor no VBP se a Demanda Final apenas do primeiro setor subisse em 10% e dos outros permanecesse constante, considerando a matriz de coeficientes que nós temos?

In [20]:
multiplicador = [1.1]
multiplicador.extend([1] * 66)

DF_2 = DF.copy() * multiplicador
DF_2.index = matriz_L.columns
VBP_2 = matriz_L.dot(DF_2)

(VBP_2/VBP)

Agricultura, inclusive o apoio à agricultura e a pós-colheita    1.056305
Pecuária, inclusive o apoio à pecuária                           1.000258
Produção florestal; pesca e aquicultura                          1.001947
Extração de carvão mineral e de minerais não metálicos           1.002366
Extração de petróleo e gás, inclusive as atividades de apoio     1.002415
                                                                   ...   
Saúde pública                                                    1.000000
Saúde privada                                                    1.000000
Atividades artísticas, criativas e de espetáculos                1.000071
Organizações associativas e outros serviços pessoais             1.000184
Serviços domésticos                                              1.000000
Length: 67, dtype: float64

Podemos ver, como esperado, que a variação no produto total é diferente em cada setor e não batem com a variação na demanda. 